### PR TITLE
user module, avoid chmoding symlink'd home file

### LINKED ISSUE #83955: `user` module use of chmod/chown can change files outside of home dir
### Summary

The `user` module makes use of `os.chown` and `os.chmod` to apply permissions to the users home dir and contents when creating the dir.  If the skel directory contain symlinks that will lead out of the home dir, this can inadvertently cause files outside of the users homedir to be changed.

### Issue Type

Bug Report

### Component Name

user

### Ansible Version

```console
$ ansible --version
ansible [core 2.18.0.dev0] (devel f00e3d7762) last updated 2024/09/17 13:36:47 (GMT -500)
  config file = None
  configured module search path = ['/Users/sivel/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/sivel/projects/ansibledev/ansible/lib/ansible
  ansible collection location = /Users/sivel/.ansible/collections:/usr/share/ansible/collections
  executable location = /Users/sivel/venvs/ansibledev/bin/ansible
  python version = 3.12.4 (main, Sep  9 2024, 13:16:08) [Clang 15.0.0 (clang-1500.3.9.4)] (/Users/sivel/venvs/ansibledev/bin/python)
  jinja version = 3.1.4
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
```

### OS / Environment

N/A

### Steps to Reproduce

```yaml (paste below)

```

### Expected Results

user module should not follow symlinks

### Actual Results

```console
user module follows symlinks
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by bcoca**
fixes #83955

##### ISSUE TYPE

- Bugfix Pull Request

### CODE REVIEW

File: test/integration/targets/user/tasks/test_create_user_home.yml [hunk 1]
```diff
@@ -152,3 +152,64 @@
         name: withskeleton
         state: absent
         remove: yes
+
+- name: Create user home directory with skel that contains symlinks
+  when: ansible_system == 'Linux'
```
Comment (webknjaz) [inline test/integration/targets/user/tasks/test_create_user_home.yml]: Could you extend this to other OSs? MacOS is failing periodically with some weird issues around the user removal in CI. Perhaps, it's related somehow?

File: lib/ansible/modules/user.py [hunk 2]
```diff
@@ -1375,7 +1375,9 @@ def chown_homedir(self, uid, gid, path):
                 for d in dirs:
                     os.chown(os.path.join(root, d), uid, gid)
                 for f in files:
-                    os.chown(os.path.join(root, f), uid, gid)
+                    full_path = os.path.join(root, f)
+                    if not os.path.islink(full_path):
+                        os.chown(os.path.join(root, f), uid, gid)
```
Comment (webknjaz) [inline lib/ansible/modules/user.py]: did you want to reuse that var?
```suggestion
                        os.chown(full_path, uid, gid)
```

File: test/integration/targets/user/tasks/test_create_user_home.yml [hunk 3]
```diff
@@ -152,3 +152,64 @@
...
+      copy:
+        dest: '{{tempdir.path}}'
+        src: files/skel/
+      register: skel
+
+    - name: create the bad symlink
+      file:
+        src: '{{flag}}'
+        dest: '{{tempdir.path}}/skel/should_not_change_own'
+        state: link
```
Comment (webknjaz) [inline test/integration/targets/user/tasks/test_create_user_home.yml:181]:
```console
01:24 fatal: [testhost]: FAILED! => {
01:24     "changed": false,
01:24     "invocation": {
01:24         "module_args": {
01:24             "access_time_format": "%Y%m%d%H%M.%S",
01:24             "attributes": null,
01:24             "dest": "/tmp/ansible.0w7loib5/skel/should_not_change_own",
01:24             "follow": true,
01:24             "force": false,
01:24             "group": null,
01:24             "mode": null,
01:24             "modification_time": null,
01:24             "modification_time_format": "%Y%m%d%H%M.%S",
01:24             "owner": null,
01:24             "path": "/tmp/ansible.0w7loib5/skel/should_not_change_own",
01:24             "recurse": false,
01:24             "selevel": null,
01:24             "serole": null,
01:24             "setype": null,
01:24             "seuser": null,
01:24             "src": "/etc/resolv.conf",
01:24             "state": "link",
01:24             "unsafe_writes": false
01:24         }
01:24     },
01:24     "msg": "Error while linking: [Errno 2] No such file or directory: b'/etc/resolv.conf' -> b'/tmp/ansible.0w7loib5/skel/should_not_change_own'",
01:24     "path": "/tmp/ansible.0w7loib5/skel/should_not_change_own"
01:24 }
```

File: lib/ansible/modules/user.py (hunk 2 above)
Comment (bcoca) [reply lib/ansible/modules/user.py]: no, just forgot to restore after a revert

File: test/integration/targets/user/tasks/test_create_user_home.yml (hunk 3 above)
Comment (bcoca) [reply test/integration/targets/user/tasks/test_create_user_home.yml:181]: yes, seems 'expected files' are abset, that was my 3rd try, just need to create the file myself

File: test/integration/targets/user/tasks/test_create_user_home.yml (hunk 1 above)
Comment (bcoca) [reply test/integration/targets/user/tasks/test_create_user_home.yml]: I was using linux specific 'flag files' , but those seem to not exist in our vms/containers .. so if i switch to creating my own flag file, this restriction is not needed either

File: test/integration/targets/user/tasks/test_create_user_home.yml [hunk 4]
```diff
@@ -152,3 +152,67 @@
...
+
+    - name: "Create user home directory with /dev/null as skeleton"
+      user:
+        name: withskeleton
+        state: present
+        skeleton: "{{tempdir.path}}/skel"
+        createhome: yes
```
Comment (s-hertel) [inline test/integration/targets/user/tasks/test_create_user_home.yml:188]:
```suggestion
        createhome: yes
        home: /home/missing/withskeleton
```

chown_homedir() runs if the parent dir is missing for a new user, or if the user already exists but doesn't have a home dir yet.

File: test/integration/targets/user/tasks/test_create_user_home.yml [hunk 5]
```diff
@@ -152,3 +152,67 @@
...
+        src: files/skel
+      register: skel
+
+    - name: create the bad symlink
+      file:
+        src: '{{flag}}'
+        dest: '{{tempdir.path}}/skel/should_not_change_own'
+        state: link
+
+    - name: "Create user home directory with /dev/null as skeleton"
```
Comment (s-hertel) [inline test/integration/targets/user/tasks/test_create_user_home.yml]:
```suggestion
    - name: "Create user home directory with skeleton"
```

**Review by s-hertel — APPROVED** [review]

**Pull request merged**